File: ios/Sources/TaigiKeyboard/Styling/ThemeImageCache.swift
// Per-process decoded-photo cache for ThemeBackgroundSurface, keyed by the theme image file name.

import SwiftUI
import UIKit

/// Which decode of a theme photo a render site needs: the stored 1280 px photo for the
/// keyboard and the editor preview, or a small downsample for cards and row thumbnails.
enum ThemeImageVariant: Hashable {
    case full
    case thumbnail

    /// Longest edge of a thumbnail decode — covers a theme card at 3×.
    static let thumbnailLongEdge: CGFloat = 480
}

/// Decodes theme photos from the App Group `ThemeImageStore` directory OFF the main actor
/// and keeps them decoded (`preparingForDisplay`, so a redraw never re-decodes the JPEG).
/// Render sites read `cached(_:_:)` (lookup only, no I/O) and fall back to `load(_:_:)`,
/// which decodes in a detached task; concurrent loads of one photo share a single decode.
/// Full photos and thumbnails are cached separately: the full cache holds at most two
/// (the keyboard's + the editor's) so the keyboard extension, under its 64 MB cap, drops
/// a previous theme's photo once another is loaded. A replaced photo gets a new file name,
/// so a stale entry is never served.
final class ThemeImageCache {
    static let shared = ThemeImageCache(store: ThemeImageStore(containerURL: SharedSettings.sharedContainerURL))
    private static let fullCountLimit = 2
    private static let fullCostLimitBytes = 14 * 1024 * 1024
    private static let thumbnailCostLimitBytes = 8 * 1024 * 1024

    private struct Key: Hashable {
        let file: String
        let variant: ThemeImageVariant
    }

    /// The one store both the host app (writer) and this cache (reader) resolve photos through.
    let store: ThemeImageStore
    private let fullCache = NSCache<NSString, UIImage>()
    private let thumbnailCache = NSCache<NSString, UIImage>()
    @MainActor private var inFlight: [Key: Task<UIImage?, Never>] = [:]

    init(store: ThemeImageStore) {
        self.store = store
        fullCache.countLimit = Self.fullCountLimit
        fullCache.totalCostLimit = Self.fullCostLimitBytes
        thumbnailCache.totalCostLimit = Self.thumbnailCostLimitBytes
    }

    /// The already-decoded photo, or nil. Never touches the disk.
    func cached(_ file: String, _ variant: ThemeImageVariant) -> UIImage? {
        cache(for: variant).object(forKey: file as NSString)
    }

    /// The decoded photo, decoding it off the main actor on a miss; nil when the file is
    /// missing / undecodable.
    @MainActor
    func load(_ file: String, _ variant: ThemeImageVariant) async -> UIImage? {
        if let cached = cached(file, variant) {
            return cached
        }
        let key = Key(file: file, variant: variant)
        if let pending = inFlight[key] {
            return await pending.value
        }
        guard let url = store.url(for: file) else { return nil }
        let task = Task.detached(priority: .userInitiated) { Self.decode(url, variant) }
        inFlight[key] = task
        let image = await task.value
        inFlight[key] = nil
        if let image {
            let cost = image.cgImage.map { $0.bytesPerRow * $0.height } ?? 0
            cache(for: variant).setObject(image, forKey: file as NSString, cost: cost)
        }
        return image
    }

    /// Drops every decoded photo (after a sweep removed files, so a deleted theme's photo
    /// does not linger in the host process).
    func removeAll() {
        fullCache.removeAllObjects()
        thumbnailCache.removeAllObjects()
    }

    private func cache(for variant: ThemeImageVariant) -> NSCache<NSString, UIImage> {
        switch variant {
        case .full: fullCache
        case .thumbnail: thumbnailCache
        }
    }

    private static func decode(_ url: URL, _ variant: ThemeImageVariant) -> UIImage? {
        switch variant {
        case .full:
            UIImage(contentsOfFile: url.path)?.preparingForDisplay()
        case .thumbnail:
            (try? Data(contentsOf: url)).flatMap { ThemeImageStore.downsampled($0, maxLongEdge: ThemeImageVariant.thumbnailLongEdge) }
        }
    }
}

/// Shows `content` with the decoded theme photo, or `placeholder` until it is decoded (or
/// when the file is missing). A cache hit renders the photo on the first frame; a miss
/// loads it off the main actor and swaps it in.
struct ThemePhotoImage<Content: View, Placeholder: View>: View {
    let file: String
    let variant: ThemeImageVariant
    @ViewBuilder let content: (UIImage) -> Content
    @ViewBuilder let placeholder: () -> Placeholder

    private struct Loaded {
        let file: String
        let variant: ThemeImageVariant
        let image: UIImage
    }

    @State private var loaded: Loaded?

    var body: some View {
        Group {
            if let image = ThemeImageCache.shared.cached(file, variant) ?? loadedImage {
                content(image)
            } else {
                placeholder()
            }
        }
        .task(id: "\(file)|\(variant)") {
            guard ThemeImageCache.shared.cached(file, variant) == nil,
                  let image = await ThemeImageCache.shared.load(file, variant) else { return }
            loaded = Loaded(file: file, variant: variant, image: image)
        }
    }

    /// The loaded image only while it still belongs to the current file / variant.
    private var loadedImage: UIImage? {
        guard let loaded, loaded.file == file, loaded.variant == variant else { return nil }
        return loaded.image
    }
}

File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/core/ThemeImageCache.kt
// Process-wide decoded theme-photo cache, keyed by the theme image file name.

package com.siansiansu.taigikeyboard.ime.core

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageDecoder
import android.util.LruCache
import androidx.annotation.MainThread
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.LocalContext
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.withContext

/** Which decode of a theme photo a render site needs. Mirrors iOS ThemeImageVariant. */
enum class ThemeImageVariant {
    /** The stored 1280 px photo — the keyboard and the editor preview. */
    FULL,

    /** A downsample to [THUMBNAIL_LONG_EDGE] — theme cards and row thumbnails. */
    THUMBNAIL,
    ;

    companion object {
        /** Longest edge of a thumbnail decode — covers a theme card at 3×. */
        const val THUMBNAIL_LONG_EDGE = 480
    }
}

/**
 * Decodes theme photos from the [ThemeImageStore] directory OFF the main thread and keeps them
 * decoded as hardware bitmaps (pixels in GPU memory, not the Java heap; both the View
 * `drawBitmap` and Compose `drawImage` paths accept it). Render sites read [cached] (lookup
 * only, no I/O) and fall back to [load], which decodes on [Dispatchers.IO]; concurrent loads of
 * one photo share a single decode. Full photos and thumbnails are cached separately; the full
 * cache holds two (the keyboard's + the editor's — the app and the IME share this
 * process). A replaced photo gets a new file name, so a stale entry is never served. One
 * instance lives on `CompositionRoot`. Mirrors iOS ThemeImageCache.
 */
class ThemeImageCache(
    val store: ThemeImageStore,
) {
    // Counted, not sized: every stored photo is capped at MAX_LONG_EDGE (~6.5 MB decoded), so two
    // entries also bound the bytes. Thumbnails vary in size and are sized by bytes.
    private val fullCache = LruCache<String, Bitmap>(FULL_COUNT_LIMIT)
    private val thumbnailCache =
        object : LruCache<String, Bitmap>(THUMBNAIL_BYTES_LIMIT) {
            override fun sizeOf(
                key: String,
                value: Bitmap,
            ): Int = value.byteCount
        }

    // Main-thread only (see [load]); the decode itself runs on IO.
    private val inFlight = HashMap<Pair<String, ThemeImageVariant>, CompletableDeferred<Bitmap?>>()

    /** The already-decoded photo, or null. Never touches the disk. */
    fun cached(
        file: String,
        variant: ThemeImageVariant,
    ): Bitmap? = cache(variant).get(file)

    /**
     * The decoded photo, decoding it on [Dispatchers.IO] on a miss; null when the file is
     * missing / undecodable. The first caller runs the decode; concurrent callers await it. The
     * decode + publish run [NonCancellable] (tens of ms) so a caller leaving composition still
     * fills the cache and releases the waiters — no cache-owned scope needed.
     */
    @MainThread
    suspend fun load(
        file: String,
        variant: ThemeImageVariant,
    ): Bitmap? {
        cached(file, variant)?.let { return it }
        val key = file to variant
        inFlight[key]?.let { return it.await() }
        val pending = CompletableDeferred<Bitmap?>()
        inFlight[key] = pending
        return withContext(NonCancellable) {
            val decoded = withContext(Dispatchers.IO) { decode(file, variant) }
            // Back on the caller's (main) dispatcher: [inFlight] stays main-confined.
            decoded?.let { cache(variant).put(file, it) }
            inFlight.remove(key)
            pending.complete(decoded)
            decoded
        }
    }

    /**
     * Removes every stored photo no theme in [themes] references and drops their decoded
     * bitmaps. Wired as the store's mutation hook and run when the editor closes.
     */
    fun sweep(themes: List<UserTheme>) {
        val referenced = themes.referencedPhotoFiles()
        store.sweep(referenced)
        for (cache in listOf(fullCache, thumbnailCache)) {
            cache
                .snapshot()
                .keys
                .filter { it !in referenced }
                .forEach { cache.remove(it) }
        }
    }

    private fun cache(variant: ThemeImageVariant): LruCache<String, Bitmap> =
        when (variant) {
            ThemeImageVariant.FULL -> fullCache
            ThemeImageVariant.THUMBNAIL -> thumbnailCache
        }

    private fun decode(
        file: String,
        variant: ThemeImageVariant,
    ): Bitmap? {
        val path = store.file(file)
        return when (variant) {
            ThemeImageVariant.FULL -> {
                val options = BitmapFactory.Options().apply { inPreferredConfig = Bitmap.Config.HARDWARE }
                BitmapFactory.decodeFile(path.path, options)
            }
            ThemeImageVariant.THUMBNAIL ->
                try {
                    ImageDecoder.decodeBitmap(ImageDecoder.createSource(path)) { decoder, info, _ ->
                        val (width, height) =
                            ThemeImageStore.targetSize(info.size.width, info.size.height, ThemeImageVariant.THUMBNAIL_LONG_EDGE)
                        decoder.setTargetSize(width, height)
                    }
                } catch (e: Exception) {
                    null
                }
        }
    }

    companion object {
        private const val FULL_COUNT_LIMIT = 2
        private const val THUMBNAIL_BYTES_LIMIT = 8 * 1024 * 1024
    }
}

/**
 * The decoded theme photo for [file] (null file, missing file, or not decoded yet → null). A
 * cache hit is returned on the first composition; a miss decodes off the main thread and
 * recomposes when it lands. Mirrors iOS ThemePhotoImage.
 */
@Composable
fun rememberThemePhoto(
    file: String?,
    variant: ThemeImageVariant,
): Bitmap? {
    val context = LocalContext.current
    val cache = remember(context) { CompositionRoot.shared(context).themeImages }
    var bitmap by remember(file, variant) { mutableStateOf(file?.let { cache.cached(it, variant) }) }
    LaunchedEffect(file, variant) {
        if (bitmap == null && file != null) bitmap = cache.load(file, variant)
    }
    return bitmap
}
